File: lib/utils/preferences_helper.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:flutter/material.dart';

class PreferencesHelper {
  static SharedPreferences? _prefs;

  /// Initialize shared preferences
  static Future<void> init() async {
    _prefs ??= await SharedPreferences.getInstance();
  }

  // -------------------- String --------------------

  static Future<bool> setString(String key, String value) async {
    await init();
    return _prefs!.setString(key, value);
  }

  static String? getString(String key) {
    return _prefs?.getString(key);
  }

  // -------------------- Bool --------------------

  static Future<bool> setBool(String key, bool value) async {
    await init();
    return _prefs!.setBool(key, value);
  }

  static bool? getBool(String key) {
    return _prefs?.getBool(key);
  }

  // -------------------- Int --------------------

  static Future<bool> setInt(String key, int value) async {
    await init();
    return _prefs!.setInt(key, value);
  }

  static int? getInt(String key) {
    return _prefs?.getInt(key);
  }

  // -------------------- Double --------------------

  static Future<bool> setDouble(String key, double value) async {
    await init();
    return _prefs!.setDouble(key, value);
  }

  static double? getDouble(String key) {
    return _prefs?.getDouble(key);
  }

  // -------------------- List<String> --------------------

  static Future<bool> setStringList(String key, List<String> value) async {
    await init();
    return _prefs!.setStringList(key, value);
  }

  static List<String>? getStringList(String key) {
    return _prefs?.getStringList(key);
  }

  // -------------------- JSON (Map) --------------------

  static Future<bool> setJson(String key, Map<String, dynamic> value) async {
    final jsonString = jsonEncode(value);
    return setString(key, jsonString);
  }

  static Map<String, dynamic>? getJson(String key) {
    final jsonString = getString(key);
    if (jsonString == null) return null;

    try {
      return jsonDecode(jsonString) as Map<String, dynamic>;
    } catch (e) {
      return null;
    }
  }
  

  // -------------------- Color --------------------

static Future<bool> setColor(String key, Color color) async {
  return setInt(key, color.value);
}

static Color? getColor(String key) {
  final colorValue = getInt(key);
  if (colorValue == null) return null;
  return Color(colorValue);
}

  // -------------------- Utilities --------------------

  static Future<bool> remove(String key) async {
    await init();
    return _prefs!.remove(key);
  }

  static Future<bool> clear() async {
    await init();
    return _prefs!.clear();
  }

  /// Optional: Log all preferences for debugging
  static void logAllPrefs() {
    if (_prefs == null) {
      print('Preferences not initialized.');
      return;
    }

    print('---- Preferences Dump ----');
    for (final key in _prefs!.getKeys()) {
      print('$key: ${_prefs!.get(key)}');
    }
    print('--------------------------');
  }


  
}

File: app/src/main/java/com/pranshulgg/weathermaster/core/utils/PreferencesHelper.kt
package com.pranshulgg.weathermaster.core.utils


import android.content.Context
import android.content.SharedPreferences
import android.graphics.Color
import android.os.Build
import androidx.annotation.RequiresApi
import androidx.core.content.edit
import org.json.JSONException
import org.json.JSONObject

object PreferencesHelper {

    private const val PREF_NAME = "app_prefs"
    private lateinit var prefs: SharedPreferences

    // -------------------- Init --------------------

    fun init(context: Context) {
        if (!::prefs.isInitialized) {
            prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
        }
    }

    // -------------------- String --------------------

    fun setString(key: String, value: String) {
        prefs.edit { putString(key, value) }
    }

    fun getString(key: String): String? = prefs.getString(key, null)

    // -------------------- Bool --------------------

    fun setBool(key: String, value: Boolean) {
        prefs.edit { putBoolean(key, value) }
    }

    fun getBool(key: String): Boolean? =
        if (prefs.contains(key)) prefs.getBoolean(key, false) else null

    // -------------------- Int --------------------

    fun setInt(key: String, value: Int) {
        prefs.edit { putInt(key, value) }
    }

    fun getInt(key: String): Int? =
        if (prefs.contains(key)) prefs.getInt(key, 0) else null

    // -------------------- Double --------------------

    fun setDouble(key: String, value: Double) {
        prefs.edit { putString(key, value.toString()) }
    }

    fun getDouble(key: String): Double? {
        val str = prefs.getString(key, null)
        return str?.toDoubleOrNull()
    }

    // -------------------- List<String> --------------------

    fun setStringList(key: String, value: List<String>) {
        prefs.edit { putStringSet(key, value.toSet()) }
    }

    fun getStringList(key: String): List<String>? =
        prefs.getStringSet(key, null)?.toList()

    // -------------------- JSON (Map) --------------------

    fun setJson(key: String, value: Map<String, Any?>) {
        val json = JSONObject(value).toString()
        setString(key, json)
    }

    fun getJson(key: String): Map<String, Any>? {
        val jsonStr = getString(key) ?: return null
        return try {
            val jsonObject = JSONObject(jsonStr)
            val map = mutableMapOf<String, Any>()
            val keys = jsonObject.keys()
            while (keys.hasNext()) {
                val k = keys.next()
                map[k] = jsonObject.get(k)
            }
            map
        } catch (e: JSONException) {
            null
        }
    }

    // -------------------- Color --------------------

    fun setColor(key: String, color: Int) {
        setInt(key, color)
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun getColor(key: String): Int? = getInt(key)?.let { Color.valueOf(it).toArgb() }

    // -------------------- Utilities --------------------

    fun remove(key: String) {
        prefs.edit { remove(key) }
    }

    fun clear() {
        prefs.edit { clear() }
    }

    fun logAllPrefs() {
        val all = prefs.all
        if (all.isEmpty()) {
            println("Preferences empty.")
        } else {
            println("---- Preferences Dump ----")
            all.forEach { (k, v) ->
                println("$k: $v")
            }
            println("--------------------------")
        }
    }
}